### ISSUE TITLE
Task: Local pining cluster test

### ISSUE THREAD

**Issue by jmozah** [ethersphere/bee#248]
The task at  hand is to test the Local Pinning feature in the cluster level with all the protocols working.

Pre-conditions:
-  A connected cluster (if kademlia is used, let it saturate for sometime)
-  Each node should be configured to have a small DB capacity (ex: 500 chunks)
- Upload a chunk in node A and download it from node B, just to see if everything is fine

Pinning Testcase 1:
- pick a node (Node A) and upload small file (ex: 10 chunks) and pin it 
- upload a larger file below the DB capacity (ex: 100 chunks) in node A
- download the small file and check if it is found and check in the pinning list API

Pinning Testcase 2:
- pick a node (Node A) and upload small file (ex: 10 chunks) and pin it 
- upload a larger file  **above** the DB capacity (ex: 1000 chunks) in node A
- download the small file and check if it is found and check in the pinning list API

Pinning Testcase 3:
- pick a node (Node A) and upload small file (ex: 10 chunks) and pin it 
- upload a larger file  **above** the DB capacity (ex: 1000 chunks) in node A
- upload many larger files above the DB capacity in node A
- download the small file before every large file upload and check if it is found and check in the pinning list API

## Acceptance criteria
- Leave the cluster for sometime (depends on our infra cost)
- Download the small file from Node A and see if it can be downloaded

Failure case:
If at any time, the smaller file cannot be downloaded, then pinning failed.

Unpinning Testcase 1:
- pick a node (Node A) and upload small file (ex: 10 chunks) and pin it 
- upload a larger file below the **above** the DB capacity (ex: 1000 chunks) in node A
- download the small file and check if it is found and check in the pinning list API
- unpin the file using the API
- upload muliple larger file below the **above** the DB capacity (ex: 1000 chunks) in node A
- download the small file and it should not be present

**Comment by jmozah** [comment]
- Add a flag during startup to specify the dbCapacity (@jmozah )
-

**Comment by svetomir** [comment]
Blocked by #259

**Comment by svetomir** [comment]
Merged into beekeeper

**Comment by Eknir** [comment]
@mortelli , could you have a look on whether this integration test covers the scenario, as discovered in global pinning in Swarm?

**Comment by Eknir** [comment]
Also make an issue to have a look at this from more theoretical perspective (based on comment @acud)

**Comment by mortelli** [comment]
so this is what we gathered together with @acud after taking some time looking at the code:

### Pinning Test Case n°3

Retrieval of the pinned file should be done only after uploading _all_ of the larger files, not after each one.

This is because each time the smaller file is retrieved, it is less likely for it to be garbage collected, since it is pushed to the back of the garbage collection queue, due to it being "used".

We could keep this test case as is, or alternatively take this idea for a 4th test case instead.

### Cluster test bugs

#### Size
It appears file sizes are taken in bytes, rather than chunks. This means that tests need to be updated so that they use the swarm chunk size when calculating the file size to be uploaded.

#### Pinned chunks
Also, it seems only the root chunks of the files to be pinned are actually pinned, rather than all of the chunks that make up the file. As far as we know, there is no "pin file" operation, but rather a "pin chunk" operation, so this is an expected (but incorrect) behavior for these test cases. 

A bug could arise if the pinned files are big, since the more chunks that make up a file, the greater the chance that they are garbage collected and the file cannot be retrieved due to only having its root chunk pinned. 

We could create a test case that pins big files instead of just small ones.

### Design flaw: race conditions

We believe the current local pinning implementation has a race condition built-in in the design. There are 2 scenarios under which we understand that content can be gc'd after being uploaded but before being pinned.

This is not so relevant for the operation of pinning content previously uploaded, as these are 2 separate operations (first upload, then pin after x time has passed) and a failure in the second one is more acceptable (i.e. attempt to pin → sorry, can't find the chunk). 
 
So, for the operation of uploading with the `pin` flag set:

#### Scenario I
1. Upload a file
2. Access it (this would put the file up for garbage collection)
3. Pin it

The file has to be big enough (or the api operation slow enough) so that it can be accessed before the pinning operation is executed, and then also garbage collected. For this, the storage should be at its limit and other parallel uploads can take place so that the GC is triggered.
 
#### Scenario II
Same as scenario I, but having the file push synced rather than accessed. This would also put the chunks up for garbage collection.

#### Testing
We can both aim at having these scenarios as beekeeper tests, as well as attempt to manually trigger them.

#### Proposal 
The solution to this problem would be to have the upload and pin operation executed in an atomic manner, rather than sequentially as 2 (internal) operations based on a single API call.

We could have a pin flag set by HTTP request headers and condense the upload and pin operations into a single one, so that on each call to the localstore which uploads the chunks, they are immediately pinned too.

Additionally, this would allow us to place the uploaded content into the GC right after it is uploaded, regardless of whether it is accessed (or maybe even synced).

**Comment by jmozah** [comment]
@mortelli  First, test case 3 never says to download the small file after every upload of the large file. It clearly says to download the file as the last step. 

Secondly, the design flaws you pointed is in the old Swarm code, why are we even talking them here.
We should say, we are yet to implement local pinning for /files, /bytes and /bzz api's.

**Comment by Eknir** [comment]
Hi @jmozah , afaik, the pinning has been mostly taken from Swarm in it's current form and, hence, design flaws there mean design flaws in Bee. 

This analysis is not personal to you or your code in any way, it is a thorough analysis of a feature which is crucial to get right.

I believe @mortelli and @acud have done a good job, inspecting the code and writing down the conclusions here. They are a good basis to improve upon local pinning. 

Constructive comments from you on the content of this ticket are very welcome, of course.

**Comment by Eknir** [comment]
I just spoke with @jmozah to clarify this.

His issue was with the seemingly infeasible test case scenario (pin file), as pinning a file is not even possible in Bee. Having said this, the research is in my opinion still valid, as it of course needs to be solved when we do pinning of a file in Bee.

In all honesty, I was not aware that the pinning operation is solely available on the `/chunks` endpoint in the current code and pinning on our other upload endpoints should be supported as well.

If the described design flaw is still possible on the `/chunks` endpoint only, I believe we should fix it right now. If it is not yet possible, I suggest we fix it when we add pinning to the other endpoints (which would enable the scenario).

W.r.t. the cluster test, this should of course be fixed asap and investigated why a faulty test still passes.

Thanks guys and let's keep peace :bird:
